integration/tsconfig-test.ts

```diff
@@ -0,0 +1,125 @@
+import { test, expect } from "@playwright/test";
+import fse from "fs-extra";
+import path from "path";
+import JSON5 from "json5";
+
+import { createFixture, json } from "./helpers/create-fixture";
+
+async function getTsConfig(projectDir: string) {
+  let tsconfigPath = path.join(projectDir, "tsconfig.json");
+  let config = await fse.readFile(tsconfigPath, "utf8");
+  return JSON5.parse(config);
+}
+
+// this is the default tsconfig.json that is shipped with `create-remix` templates
+const DEFAULT_CONFIG = {
+  include: ["remix.env.d.ts", "**/*.ts", "**/*.tsx"],
+  compilerOptions: {
+    allowJs: true,
+    esModuleInterop: true,
+    forceConsistentCasingInFileNames: true,
+    isolatedModules: true,
+    jsx: "react-jsx",
+    lib: ["DOM", "DOM.Iterable", "ES2019"],
+    moduleResolution: "node",
+    noEmit: true,
+    resolveJsonModule: true,
+    strict: true,
+    target: "ES2019",
+    baseUrl: ".",
+    paths: {
+      "~/*": ["./app/*"],
+    },
+  },
+};
+
+// besides baseUrl, due to an upstream issue in tsconfig-paths
+// https://github.com/dividab/tsconfig-paths/pull/208
+test("should output default tsconfig if file is empty", async () => {
+  let fixture = await createFixture({
+    files: {
+      "tsconfig.json": json({ compilerOptions: { baseUrl: "." } }),
+    },
+  });
+
+  let tsconfig = await getTsConfig(fixture.projectDir);
+  expect(tsconfig).toEqual(DEFAULT_CONFIG);
+});
+
+test("should add/update mandatory config", async () => {
+  let config = {
+    ...DEFAULT_CONFIG,
+    compilerOptions: {
+      ...DEFAULT_CONFIG.compilerOptions,
+      isolatedModules: false, // true is required by esbuild
+    },
+  };
+  delete config.compilerOptions.moduleResolution; // this is required by esbuild
+  let fixture = await createFixture({
+    files: { "tsconfig.json": json(config) },
+  });
+
+  let tsconfig = await getTsConfig(fixture.projectDir);
+  expect(tsconfig).toEqual(DEFAULT_CONFIG);
+});
+
+test("shouldn't change suggested config if set", async () => {
+  let config = {
+    ...DEFAULT_CONFIG,
+    compilerOptions: {
+      ...DEFAULT_CONFIG.compilerOptions,
+      strict: false,
+    },
+  };
+
+  let fixture = await createFixture({
+    files: {
+      "tsconfig.json": json(config),
+    },
+  });
+
+  let tsconfig = await getTsConfig(fixture.projectDir);
+  expect(tsconfig).toEqual(config);
+});
+
+test("allows for `extends` in tsconfig", async () => {
+  let config = {
+    extends: "./tsconfig.base.json",
+  };
+
+  let baseConfig = {
+    compilerOptions: {
+      allowJs: true,
+      baseUrl: ".",
+    },
+  };
+
+  let fixture = await createFixture({
+    files: {
+      "tsconfig.json": json(config),
+      "tsconfig.base.json": json(baseConfig),
+    },
+  });
+
+  let tsconfig = await getTsConfig(fixture.projectDir);
+  // our base config only sets a few options, so our local config should fill in the missing ones
```

Comment:
It should be obvious from the code which were "the missing ones" from the way this code is written. Is this just basically "the config the project had before plus some filled in properties?" If that's the case, then I should be able to tell just by reading the test code.

Reply:
in this case its the `DEFAULT_CONFIG` minus https://github.com/remix-run/remix/pull/2961/files/22e522bf19dfe9bfc60e9a365fa9223491c302de#diff-1d51fc796c50fdfde6dfe89b8fb099b6c8b25edba2b4e7da0e861a8ea8e35e1eR91-R94 as that was defined in a different tsconfig which was extended – so we could spread the default config in and delete those 2 options by hand instead of a big object

Reply:
I see. Thanks for explaining that and updating the code. So if people have `allowJs` and `baseUrl` set, then we remove those for them? When do we make this change again? I'm still unsure why we want to do that.

Reply:
so in this test, we're verifying that their tsconfig and any they're extending are merged properly, we're not removing any properties. 

this will run during the initial build of remix dev/watch and remix build

Reply:
Oh I see. So this test is just saying: "If their main config doesn't have these properties, we also check any extends to make sure those properties are in there so we don't have to add those properties manually"?

Reply:
yup! if the extend has it, we can rely on that

Reply:
> this will run during the initial build of remix dev/watch and remix build

Ok, where are we running the build in this test to make sure the code under test is actually being run?

Also, I guess I'm a bit confused still... So first we create a `tsconfig.json` with the following contents:

```json
{
  extends: "./tsconfig.base.json",
};
```

Then we create the fixture.

Then we get the file and make sure it's contents are equal to the default config. But they shouldn't be right? Because we didn't modify them.

Reply:
We talked offline and I get now. This is good 👍